Retry starting the display server

Vogtinator opened the pull request:
Even if the CanGraphical property of a Seat is true, it's possible that it's
still too early for X to start, as it might need some driver or device which
isn't present yet.

Fixes #1316

plfiorini commented:
Can't we do something async? DisplayServer emtis a started() signal.

tsdgeos commented:
Unfortunately this doesn't fix things for me, it even breaks them more.

For some reason i can't even switch to VT2 anymore

albert-astals-cid-kdab commented:
The journalctl log is super weird though, i can't understnad why it has two attempt 0 
[with_patch.txt](https://github.com/sddm/sddm/files/5533723/with_patch.txt)

albert-astals-cid-kdab commented:
This seems to work better https://github.com/tsdgeos/sddm/commit/3eb5c0ece63b6942b7d7fafb4b479018cda3f9d7 for me.

No idea how dangerous is to spin an event loop there

Vogtinator commented:
I see what happens. The `QProcess` from the failed X server sent a signal that it exited, but no event loop is running so it's only delivered after the next attempt succeeded and it returns to the event loop. Unfortunately this signal causes X to be stopped again...

So what needs to be done is to wait until the old process finished and got cleaned up, and only then the next attempt can be made.

Vogtinator commented:
`XorgDisplayServer` doesn't really handle the case where `start()` fails and is later called again. It doesn't clean up properly and also doesn't emit a signal when the old process is gone and can be started again. So currently this waits 2s and hopes for the best.

Any change there would conflict with https://github.com/sddm/sddm/pull/1312, so I'd either wait until that is merged or add that to the commits here.

It would be ideal if when `start()` fails, it blocks until the `QProcess` is deleted, then this would "just work".

tsdgeos commented:
I can confirm this seems to fix it, it aint pretty but it isn't mega terribly "slow" either

tsdgeos commented:
Now, would it make sense to try to talk to the systemd people about this "wait for all cards to be ready before trying to start X"?

Vogtinator commented:
> Now, would it make sense to try to talk to the systemd people about this "wait for all cards to be ready before trying to start X"?

That's not possible, as it's not known in advance which cards are expected but not initialized yet. It's only known after the fact. Someone might plug in one over USB 10 minutes after booting, for instance.

davidedmundson reviewed (approved):
Thanks for following up on this

tsdgeos commented:
Can we get this landed and 0.19.1 released? 

Not being able to use the software at all seems like a huge regression to me.

tsdgeos commented:
Ping?

Vogtinator commented:
> XorgDisplayServer doesn't really handle the case where start() fails and is later called again. It doesn't clean up properly

I changed it so that when `XorgDisplayServer::start` fails, it makes sure that the process is stopped/stopping when it returns and added a check to ensure that an early call to `start` before the previous `process` finished does not happen.

tsdgeos commented:
ok, i can confirm that this branch still fixes the issue for me.
